streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -119,6 +123,7 @@ public void put(final Bytes key,
         validateStoreOpen();
         lock.writeLock().lock();
         try {
+            validateStoreOpen();
```

Comment:
Why do we  call this a second time? Also, it seems to be a very generic pattern for all method calls -- can we extract this in some private helper method ?

Reply:
We need to check again after we grab the lock, otherwise the store might get closed after we check but before we grab the lock. Once we get the lock, we're guaranteed that this block is serialized wrt close(). But we can still check beforehand to avoid grabbing the lock if it is closed.

Reply:
We can extract a helper, but maybe we can do _that_ in a follow-up at least, since the concurrency controls needs to be standardized across all the caching stores anyway.

Reply:
Ack.